mentat/conversation.py

```diff
@@ -269,3 +269,76 @@ async def get_model_response(self) -> list[FileEdit]:
         )
         self.add_model_message(parsed_llm_response.full_response, messages_snapshot)
         return parsed_llm_response.file_edits
+
+    def remaining_context(self) -> int | None:
+        ctx = SESSION_CONTEXT.get()
+        max_context = model_context_size(ctx.config.model)
+        if max_context is None:
+            return None
+
+        return max_context - conversation_tokens(self.get_messages(), ctx.config.model)
+
+    def can_add_to_context(self, message: str) -> bool:
+        """
+        Whether or not the model has enough context remaining to add this message.
+        Will take token buffer into account and uses full_message=True.
+        """
+        ctx = SESSION_CONTEXT.get()
+
+        remaining_context = self.remaining_context()
+        return (
+            remaining_context is not None
+            and remaining_context
+            - count_tokens(message, ctx.config.model, full_message=True)
+            - ctx.config.token_buffer
+            > 0
+        )
+
+    async def run_command(self, command: list[str]) -> bool:
+        """
+        Runs a command and, if there is room, adds the output to the conversation under the 'system' role.
+        """
+        ctx = SESSION_CONTEXT.get()
+        ctx.stream.send("Command output:", color="cyan")
+
+        process = subprocess.Popen(
+            command,
+            cwd=ctx.cwd,
+            text=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+        )
+        output = list[str]()
+        while True:
+            if process.stdout is None:
+                break
+            line = process.stdout.readline()
+            if not line:
+                break
+            output.append(line)
+            ctx.stream.send(line, end="")
+            # This gives control back to the asyncio event loop so we can actually print what we sent
+            # Unfortunately asyncio.sleep(0) won't work https://stackoverflow.com/a/74505785
+            # Note: if subprocess doesn't flush, output can't and won't be streamed.
+            await asyncio.sleep(0.01)
+        output = "".join(output)
+        message = (
+            f"The user ran this command:\n{' '.join(command)}\nAnd recieved this"
+            f" output:\n{output}"
+        )
+
+        if self.can_add_to_context(message):
+            self.add_message(
+                ChatCompletionSystemMessageParam(role="system", content=message)
+            )
+            ctx.stream.send(
+                "Successfully added command output to model context.", color="cyan"
+            )
+            return True
+        else:
+            ctx.stream.send(
+                "Not enough tokens remaining in model's context to add command output"
+                " to model context.",
+                color="light_red",
```

Comment:
I don't want to add too many knobs but it might be worth thinking about UX for putting a subset into context? Two ideas:
1. When output exceeds 500 tokens prompt: `Command output was X tokens: (A)dd all/add (F)irst 500 tokens/add (L)ast 500 tokens/add (N)one`.
2. When output exceeds 500 tokens prompt: `Command output was X tokens: (A)dd all to context/(D)on't add to context/Or enter a prompt describing what subset to put into context and a summarized version of the output will be put into context`.

For version one I think this behavior is fine.

Reply:
Hmm, maybe but I think this is adding too much complexity for not much gain. We can think about it later either way

Reply:
I think it's more important than it seems because perhaps the most useful thing to do with this is run tests which can often be quite verbose on failures especially if there are many and it should be easy for the user to specify the range with relevant information. But agree we can handle it later.